pkg/dataobj/index/index_builder.go

```diff
@@ -0,0 +1,510 @@
+	EnabledTenantIDs       flagext.StringSliceCSV `yaml:"enabled_tenant_ids" experimental:"true"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("dataobj-index-builder.", f)
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	cfg.BuilderConfig.RegisterFlagsWithPrefix(prefix, f)
+	f.IntVar(&cfg.EventsPerIndex, prefix+"events-per-index", 32, "Experimental: The number of events to batch before building an index")
+	f.StringVar(&cfg.IndexStoragePrefix, prefix+"storage-prefix", "index/v0/", "Experimental: A prefix to use for storing indexes in object storage. Used to separate the metastore & index files during initial testing.")
+	f.Var(&cfg.EnabledTenantIDs, prefix+"enabled-tenant-ids", "Experimental: A list of tenant IDs to enable index building for. If empty, all tenants will be enabled.")
+}
+
+type downloadedObject struct {
+	event       metastore.ObjectWrittenEvent
+	objectBytes *[]byte
+	err         error
+}
+
+const indexEventTopic = "loki.metastore-events"
+const indexConsumerGroup = "metastore-event-reader"
+
+type Builder struct {
+	services.Service
+
+	cfg Config
+
+	// Kafka client and topic/partition info
+	client *kgo.Client
+	topic  string
+
+	// Processing pipeline
+	downloadQueue     chan metastore.ObjectWrittenEvent
+	downloadedObjects chan downloadedObject
+	builder           *indexobj.Builder
+
+	bufferedEvents map[string][]metastore.ObjectWrittenEvent
+
+	// Builder initialization
+	builderCfg  indexobj.BuilderConfig
+	bucket      objstore.Bucket
+	flushBuffer *bytes.Buffer
+
+	// Metrics
+	metrics *indexBuilderMetrics
+
+	// Control and coordination
+	ctx        context.Context
+	cancel     context.CancelFunc
+	wg         sync.WaitGroup
+	logger     log.Logger
+	builderMtx sync.Mutex
+}
+
+func NewIndexBuilder(
+	cfg Config,
+	kafkaCfg kafka.Config,
+	logger log.Logger,
+	instanceID string,
+	bucket objstore.Bucket,
+	reg prometheus.Registerer,
+) (*Builder, error) {
+	kafkaCfg.AutoCreateTopicEnabled = true
+	kafkaCfg.AutoCreateTopicDefaultPartitions = 64
+	eventConsumerClient, err := client.NewReaderClient(
+		"index_builder",
+		kafkaCfg,
+		logger,
+		reg,
+		kgo.ConsumeTopics(indexEventTopic),
+		kgo.InstanceID(instanceID),
+		kgo.SessionTimeout(3*time.Minute),
+		kgo.ConsumerGroup(indexConsumerGroup),
+		kgo.RebalanceTimeout(5*time.Minute),
+		kgo.DisableAutoCommit(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kafka consumer client: %w", err)
+	}
+
+	reg = prometheus.WrapRegistererWith(prometheus.Labels{
+		"topic":     indexEventTopic,
+		"component": "index_builder",
+	}, reg)
+
+	metrics := newIndexBuilderMetrics()
+	if err := metrics.register(reg); err != nil {
+		return nil, fmt.Errorf("failed to register metrics for index builder: %w", err)
+	}
+
+	builder, err := indexobj.NewBuilder(cfg.BuilderConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create builder: %w", err)
```

Comment:
nit: be consistent with the previous error message

```suggestion
		return nil, fmt.Errorf("failed to create index builder: %w", err)
```